### PR TITLE
Ravil/gpu integration

### PR DESCRIPTION
**Pull request by ravil-mobile**
Extended to generate GEMM and CopyAddScale kernels for GPU-architectures via GemmForge back-end

### CODE REVIEW

File: yateto/gemm_configuration.py [hunk 1]
```diff
@@ -139,10 +139,18 @@ def __init__(self, arch, threshold: int = 128):
...
-  def supported(self, m, n, k, sparseA, sparseB, transA, transB, alpha, beta):
-    return self._archSupported() and not (sparseA and sparseB) and (not transA and not transB) and alpha == 1.0 and beta in [0.0, 1.0]
+    if self._arch.name.lower() in supported_set:
+      return True
+    else:
+      if self._arch.host_name and self._arch.host_name.lower() in supported_set:
+        return True
+      else:
+        return False
+
```
Comment (krenzland) [inline yateto/gemm_configuration.py:151]: should be return self._arch.host_name and self._arch.host_name.lower() in supported_set

File: yateto/gemm_configuration.py [hunk 2]
```diff
@@ -159,10 +167,18 @@ def __init__(self, arch, threshold: int = 128):
...
   def _archSupported(self):
-    return self._arch.name.lower() in {'thunderx2t99', 'knl', 'skx'}
+    supported_set = {'thunderx2t99', 'knl', 'skx'}
+    if self._arch.name.lower() in supported_set:
+      return True
+    else:
+      if self._arch.host_name and self._arch.host_name.lower() in supported_set:
+        return True
+      else:
+        return False
```
Comment (krenzland) [inline yateto/gemm_configuration.py:177]: Should be return self._arch.host_name and self._arch.host_name.lower() in supported_set

File: yateto/gemm_configuration.py [hunk 3]
```diff
@@ -178,15 +194,41 @@ def preference(self, m, n, k, sparseA, sparseB, transA, transB, alpha, beta):
...
+    self._threshold = threshold
+
+  def _is_arch_supported(self):
+    return self._arch.name.lower() in {'nvidia'}
+
+  def supported(self, m, n, k, sparseA, sparseB, transA, transB, alpha, beta, platform):
+    if self._is_arch_supported():
+      if not (sparseA and sparseB):
+        if platform == 'gpu':
+          return True
```
Comment (krenzland) [inline yateto/gemm_configuration.py:210]: should be return self._is_arch_supported() and not (sparseA and sparseB) and platform == 'gpu'
But: do you actually support the case where one matrix is sparse?

File: yateto/generator.py [hunk 4]
```diff
@@ -341,7 +367,7 @@ def unit_test_body(cpp, testFramework):
     print('Calling external code generators...')
     with Cpp(fRoutines.h) as header:
       with header.HeaderGuard(self._headerGuardName(namespace, self.ROUTINES_FILE_NAME)):
-        cache.generate(header, fRoutines.cpp)
+        cache.generate(header, fRoutines.cpp, fGpulikeRoutines.cpp)
```
Comment (krenzland) [inline yateto/generator.py:370]: Do we want to generate the gpulike routines always? Even when not using a gpu?

File: yateto/input.py [hunk 5]
```diff
@@ -22,21 +22,51 @@ def __transposeMatrix(matrix):
 
 def __processMatrix(name, rows, columns, entries, clones, transpose, alignStride, namespace=None):
   matrix = dict()
+  is_values_booleans = False
```
Comment (krenzland) [inline yateto/input.py:25]: should be are_values_boolean or are_values_booleans

File: yateto/input.py [hunk 6]
```diff
@@ -22,21 +22,51 @@ def __transposeMatrix(matrix):
...
     row = int(entry[0])-1
     col = int(entry[1])-1
+
+    # allocate a matrix element inside of a table
     matrix[(row, col)] = entry[2]
 
+    # if all entries given as booleans it means that the matrix has the same
+    # sparsity pattern for all mesh elements but different values for each
+    # In other words, the matrix is not constant
+    is_values_booleans = isinstance(entry[2], bool)
```
Comment (krenzland) [inline yateto/input.py:40]: Shouldn't this be is_values_booleans = is_values_booleans and isinstance(entry[2], bool)?

File: yateto/input.py [hunk 7]
```diff
@@ -22,21 +22,51 @@ def __transposeMatrix(matrix):
...
     mtx = __transposeMatrix(matrix) if transpose(name) else matrix
+
+    # adjust layout description in case if a given matrix is a vector
     if len(shape) == 1:
       mtx = {(i[0],): val for i,val in mtx.items()}
-    matrices[name] = Tensor(name, shape, mtx, alignStride=alignStride(name), namespace=namespace)
+
+    # Create an tensor(matrix) using the matrix description and append the hash table
+    matrices[name] = Tensor(name=name,
+                            shape=shape,
```
Comment (krenzland) [inline yateto/input.py:65]: Question: Can it happen that name is not unique? In this case, it would maybe be better to throw an exception then?

**Review by krenzland — CHANGES_REQUESTED** [review]
Looks quite good overall, I've added some comments regarding confusing names/naming and some small possible logic errors for edge-cases.

**Comment by krenzland** [comment]
Oh, and another question: There's a Jenkinfile included - is this actually used somewhere?

File: yateto/gemm_configuration.py (hunk 3 above)
Comment (ravil-mobile) [reply yateto/gemm_configuration.py:210]: Here is a mistake in de Morgan's Law.
```
 if not (sparseA and sparseB):
```
It should be:
```
 if not (sparseA or sparseB):
```


> should be return self._is_arch_supported() and ... and ...

I know what you mean, I don't like it. It is difficult to read and understand. I would agree to merge two last inner if-statements
```
if self._is_arch_supported():
  if platform == 'gpu' and not (sparseA or sparseB):
    return True
```

File: yateto/generator.py (hunk 4 above)
Comment (ravil-mobile) [reply yateto/generator.py:370]: We generate **subroutines.cpp** and **subroutines.h** even if we don't use libsxmm. In this case, the files are just empty. I followed the same idea. If the user don't use GPU then **fGpulikeRoutines.cpp** is going to be empty. I cannot see  a problem here

**Comment by uphoffc** [comment]
Hi,

thanks for your contribution.

Before going into details, I have a few questions:
- What is the purpose of TmpMemManagerT?
- What is "sub_name" and "host_name"? And what is "compute_ident", "compute_sub_arch", "host_ident"?
- To what concept does "is_compute_constant" refer?

And a few high-level comments:
- I don't think gemmforge should be a hard dependency. You can use importlib to check whether packages are available: https://github.com/SeisSol/yateto/blob/master/yateto/ast/visitor.py#L13
- Please don't do that: "temp_variable_name = re.compile(r'_tmp*')". I see that you want to treat temporary tensors differently, but this information should be part of the CFG and not deduced with a pattern matching hack.
- I don't see why "platform" is not part of arch.
- Examples don't compile anymore.

And a more involved comment:
All the snippets like "if self._platform == 'cpu':" and "isinstance(Gpu...)" are indications of bad design and missing abstractions. Clearly, the design of the whole "codegen" folder was designed rather poorly by me, but that doesn't mean that we should continue like this. It's not only the gpu backend but it's also difficult to write an interface to another language (e.g. Fortran, C, different C++ standards). I think we need to introduce another IR.

Cheers,
Carsten

**Comment by ravil-mobile** [comment]
> Hi,
> 
> thanks for your contribution.
> 
> Before going into details, I have a few questions:
> 
> * What is the purpose of TmpMemManagerT?
> * What is "sub_name" and "host_name"? And what is "compute_ident", "compute_sub_arch", "host_ident"?
> * To what concept does "is_compute_constant" refer?
> 
> And a few high-level comments:
> 
> * I don't think gemmforge should be a hard dependency. You can use importlib to check whether packages are available: https://github.com/SeisSol/yateto/blob/master/yateto/ast/visitor.py#L13
> * Please don't do that: "temp_variable_name = re.compile(r'_tmp*')". I see that you want to treat temporary tensors differently, but this information should be part of the CFG and not deduced with a pattern matching hack.
> * I don't see why "platform" is not part of arch.
> * Examples don't compile anymore.
> 
> And a more involved comment:
> All the snippets like "if self._platform == 'cpu':" and "isinstance(Gpu...)" are indications of bad design and missing abstractions. Clearly, the design of the whole "codegen" folder was designed rather poorly by me, but that doesn't mean that we should continue like this. It's not only the gpu backend but it's also difficult to write an interface to another language (e.g. Fortran, C, different C++ standards). I think we need to introduce another IR.
> 
> Cheers,
> Carsten

Hi, 

1) TmpMemManagerT allows you to provide memory for temporary variables from SeisSol. Usually, Yateto allocates chunks of temp. memory on the stack and it works fine but it doesn’t work in case of GPUs. For GPUs, one needs to allocate a chunk of memory on the GPU heap which, in a naive way, slows down the overall time because memory allocation is delegated to the device driver. 
	To address this problem, we allocate a huge chunk of GPU memory at the beginning of SeisSol execution and handle it as a stack. Before launching a Yateto’s kernel the user takes a necessary amount of data from the GPU stack and passes it to TmpMemManagerT.  TmpMemManagerT helps to distribute the provided memory between tmp. variables. I didn’t want to use SeisSol GPU stack data structure in Yateto and I decided to abstract this problem by introducing TmpMemManagerT. 
	In general,  TmpMemManagerT can be useful even for a CPU execution when tensors become huge and there is a need to use the heap. Instead of using malloc and free (in a way how it is currently done in yateto) the user can create a similar data structure as the GPU stack but on CPU. All in all, it is supposed to reduce a number of sys-calls in such a scenario.



2) There is a necessity to compile some kernels for both CPU or/and GPU: 1) it is better leave initialization and output parts of SeiSol, where Yateto is used, on CPU 2) some compute-kernels are not trivial to adapt for GPU computing without code refactoring (e.g., point sources). In any case, padding of data structures should be imposed by GPU alignment.

	Traditionally heterogeneous computing comes with definitions of off-loading unit (Nvidia calls it as a *device*, OpenCL comes with an even more abstract name) and host unit. I didn’t want to stick to Nvidia’s definitions and I decided to name device as a compute unit. I didn’t want to propagate *if-else statements* throughout the entire code base always cheeking whether an instance of Yateto Architecture is GPU or not (it would be really ugly). I decided to make it a bit abstract and stick to another convention: all performance decisions are done based on alignment and capabilities of compute unit. If a code is supposed to get generated for CPU then compute name will be the same as host name. In case of GPUs,  these names will be different. 

	GPUs can come from different vendors. For example: AMD, Nvidia, Intel. Each vendor names differently their Streaming-Multiprocessors/SIMD-Engines/only-God-knows-how-INTEL-will-name-it (e.g. sm_60, sm_61, sm_70, etc for Nvidia). Pure names as I gave in the example are difficult to digest for the user and this is the reason why I came to the definition of compute_sub_arch (sub_name). Yateto only needs to know a vendor name while GemmForge needs to know the exact sub-architecture to optimize block-thread distribution. In short, in case of CPU host_name == compute_name and sub_name is None. If GPU then compute_name is vendor, host_name is a name of CPU (as it was before), sub_name a specific model of GPU’s SM.


3)  is_compute_constant – it refers to tensors which are known at compile-time. It means that we know both sparsity patterns and numerical values (e.g., mass and stiffness matrices). It is important for the GPU implementation because in this particular case GemmForge will generate one of two possible GEMM kernels: one-to-many or many-to-one. This leads to two benefits: the unique tensor (denoted as one) is going to be cached in each L1 GPU cache which improves arithmetic intensity; the user doesn’t have to provide an array of pointers for the unique matrix i.e., where each element should point to the same matrix. This is super convenient from SeisSol’s programming perspective


comments:
1) Thanks. I was looking for something like this
2) Yep, it is temporary solution. I will try to find a solution for this.
3) platform, is a maker for a kernel which tells you whether the current tensor expression should be generated for GPU and CPU. Note, not all tensor expression are generated for GPU (initialization, output, point sources are computed with CPU always)
4) Thanks. I forgot to check it. It will be fixed soon.

involved comment:
1) I think that Yateto does two things that can be decoupled. 1) Generating interfaces between generated kernels and SeisSol (kernel.h, tensor.h/cpp, init.h/cpp, etc) 2) generating source code for tensor expressions (kernel.cpp, subroutine.h/cpp). One solution can be the Bridge pattern because there is 2 orthogonal abstractions: C++/Fortran and CPU/GPU

2) I agree about another IR. There are two optimization steps that are machine independent: EQSPP and strength reduction. AST after these two steps can be a good candidate for IR which can be passed to a specific code generator. Each code generator should implement its own CopyAddScale, GEMM, LoG, Product, IndexSum

3) I guess we came to a new set of requirements and we need to consider the source code refactoring at some point. That is the reason why I introduced tests and created Jenkins pipeline for on-demand testing. Maybe it will be useful at some point.


Thanks for your comments!

**Comment by uphoffc** [comment]
I think the standard term for your TmpMemManagerT is "linear allocator". See e.g.

https://nfrechette.github.io/2015/05/21/linear_allocator/
https://blog.molecular-matters.com/2012/08/14/memory-allocation-strategies-a-linear-allocator/
https://github.com/mtrebi/memory-allocators

A linear allocator is indeed useful for purposes other than GPU programming (I use it in another project, too: https://github.com/TEAR-ERC/tandem/blob/seas/src/util/LinearAllocator.h).

I think we could replace stack allocation by a linear allocator also for the CPU variant, then there is no need to distinguish in the code generation phase.

Please call it linear allocator to make the concept clear from the name. 

Going to read the rest tomorrow :-)

**Comment by krenzland** [comment]
> Please call it linear allocator to make the concept clear from the name.

Haven't read the rest but a comment on this:
I always thought these things were called "bump allocators". Not sure which term is more common (or even if both describe the same concept).

**Comment by ravil-mobile** [comment]
I was thinking yesterday. I guess it is better to rename **_platform_** to **_target_**. What do you think?

File: tests/code-gen/generator.py [hunk 8]
```diff
@@ -60,75 +60,4 @@
...
-          cpp('trash[i] += trash[i];')
-
-with Cpp(os.path.join(outDir, 'performance.cpp')) as cpp:
-  cpp.includeSys('cstdlib')
-  cpp.includeSys('cstdio')
-  cpp.includeSys('cmath')
-  cpp.include('kernel.h')
-  cpp.include('tensor.h')
-  cpp.include('Stopwatch.h')
-  cpp.include('Util.h')
```
[the remark this answers is not in the document]
Comment (ravil-mobile) [reply tests/code-gen/generator.py:84]: I copied this code from examples to "unit-test" generator. This code is used to generate performance tests which are unnecessary for "unit-testing" purposes. The tests in **tests** folder should only verify that the optimized code produces the same numerical results as the naive implementation

File: yateto/codegen/common.py [hunk 9]
```diff
@@ -3,24 +3,28 @@
 from ..ast.log import splitByDistance
 
 class TensorDescription(object):
-  def __init__(self, name, memoryLayout, eqspp):
+  def __init__(self, name, memoryLayout, eqspp, is_compute_constant=False):
```
[the remark this answers is not in the document]
Comment (ravil-mobile) [reply yateto/codegen/common.py:6]: *is_compute_constant == True* refers to tensors which are known at compile-time. It means that we know both sparsity patterns and numerical values (e.g., mass and stiffness matrices). It is necessary for code generation for GPU.

File: yateto/codegen/copyscaleadd/csa_gen.py [hunk 10]
```diff
@@ -0,0 +1,176 @@
+from ..common import *
+from ..cache import RoutineGenerator, GpuRoutineGenerator
+from gemmforge import DenseMatrix, CsaGenerator, GenerationError
+from gemmforge import arch as GemmForgeArch
+import re
+
+from yateto.type import Tensor
+
+
+class CsaGen(object):
```
[the remark this answers is not in the document]
Comment (ravil-mobile) [reply yateto/codegen/copyscaleadd/csa_gen.py:10]: CopyScaleAdd

File: yateto/codegen/log/generic.py [hunk 11]
```diff
@@ -74,9 +80,9 @@ def generate(self, cpp, routineCache, gemm_cfg):
     Ceqspp = self._reduce(d.result, C, CmemLayout)
 
     gemmDescr = gemm.Description(
-      leftTerm = TensorDescription(innerAname, AmemLayout, Aeqspp),
-      rightTerm = TensorDescription(innerBname, BmemLayout, Beqspp),
-      result = TensorDescription(innerCname, CmemLayout, Ceqspp),
+      leftTerm=TensorDescription(innerAname, AmemLayout, Aeqspp, d.leftTerm.is_compute_constant),
+      rightTerm=TensorDescription(innerBname, BmemLayout, Beqspp, d.rightTerm.is_compute_constant),
+      result=TensorDescription(innerCname, CmemLayout, Ceqspp, d.result.is_compute_constant),
```
[the remark this answers is not in the document]
Comment (ravil-mobile) [reply yateto/codegen/log/generic.py:85]: These are named parameters of *Description constructor class*. According PEP8 one doesn't need to use white-spaces here

From PEP8:
*Don't use spaces around the = sign when used to indicate a keyword argument, or when used to indicate a default value for an unannotated function paramete*

```
# Correct:
def complex(real, imag=0.0):
    return magic(r=real, i=imag)

# Wrong:
def complex(real, imag = 0.0):
    return magic(r = real, i = imag)
```

The built-in linter in Pycharm emits warnings at this piece of code. To resolve it, I had to remove white-spaces

[the remark this answers is not in the document]
Comment (ravil-mobile) [reply yateto/codegen/log/generic.py:85]: I can revert it - i.e., add white-spaces - because there are many such warnings in Yateto. If I do this the code will be kinda uniformed but it won't be a proper python style

File: yateto/codegen/gemm/gemmgen.py [hunk 12]
```diff
@@ -74,13 +84,94 @@ def generate(self, cpp, routineCache):
       flops = 2 * m.size() * n.size() * k.size()
     
     if isinstance(self._gemm_cfg, BLASlike):
```
[the remark this answers is not in the document]
Comment (ravil-mobile) [reply yateto/codegen/gemm/gemmgen.py:86]: There is a way which came to my mind. It is based on double-dispatch (something like the Visitor pattern). Each *if-block* goes to a separate *generate* function. Selection of a specific *generate* function is delegated of sub-classes of *GemmTool* class (which self._gemm_cfg belongs to). Of course something needs to be done with the code above that you mentioned.

I am still not sure whether the double-dispatch is a good idea. What do you think? I would suggest to tackle this problem after the merge.

[the remark this answers is not in the document]
Comment (uphoffc) [reply yateto/codegen/gemm/gemmgen.py:86]: Double-dispatch is fine here. This is wrong, btw: num_rows=d.leftTerm.memoryLayout._bbox[0].stop,

File: yateto/arch.py [hunk 13]
```diff
@@ -40,8 +40,11 @@
 from .memory import DenseMemoryLayout
 
 class Architecture(object):
-  def __init__(self, name, precision, alignment, enablePrefetch=False):
+  def __init__(self, name, sub_name, precision, alignment, enablePrefetch=False, host_name=None):
```
[the remark this answers is not in the document]
Comment (ravil-mobile) [reply yateto/arch.py:43]: No, it is not a networking package. host in this context means a name of the host system using cuda jargon

File: yateto/codegen/copyscaleadd/csa_gen.py [hunk 14]
```diff
@@ -0,0 +1,176 @@
...
+
+    if d.beta == 1.0 and alpha == -1.0:
+      alpha = 1.0
+      flop += 1
+
+    elif d.beta == 1.0:
+      flop += 1
+
+    elif d.beta != 0.0:
+      raise NotImplementedError
```
[the remark this answers is not in the document]
Comment (ravil-mobile) [reply yateto/codegen/copyscaleadd/csa_gen.py:83]: From python reference manual:

```
raise_stmt:     "raise" [expression ["," expression ["," expression]]]
```


*If no expressions are present, raise re-raises the last expression that was raised in the current scope.*

*Otherwise, raise evaluates its first expression, which must yield a string, class, or instance object ... If the first object is a class or instance, it raises the exception identified by the class of the instance determined in the previous step, with the instance as its parameter.*

In short, you can raise either a class or an object. Anyway, a trace in case of an error will show you where a processes stops and it is more then enough for debugging

File: yateto/codegen/copyscaleadd/csa_gen.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (ravil-mobile) [reply yateto/codegen/copyscaleadd/csa_gen.py:9]: I added a class comment below the class declaration.

File: yateto/codegen/gemm/gemmgen.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (ravil-mobile) [reply yateto/codegen/gemm/gemmgen.py:97]: Gemmforge uses a different format for describing matrices i.e., num_rows is total number of rows including padding. The same is for num_columns. There is check in yateto which ensures that all tensor operations are for tensors with ranks less or equal to 2 in case of GPU computing.

The next big release will learn gemmforge to operate on tensors and combine multiple tensors sub-expressions in one GPU kernel

File: yateto/codegen/cache.py [hunk 15]
```diff
@@ -18,11 +22,18 @@ def addRoutine(self, name, generator):
...
+            generator.header(cpp)
 
     for name, generator in self._routines.items():
-      declaration = generator(name, cppFileName)
-      header(declaration)
+      if isinstance(generator, GpuRoutineGenerator):
+        declaration = generator(name, gpuFileName)
+      else:
+        declaration = generator(name, cppFileName)
+      header(declaration)
```
Comment (krenzland) [inline yateto/codegen/cache.py:39]: Not a fan of this pattern - this is bad OOP design. No need to change now, but please keep this in mind in the future.

File: yateto/codegen/copyscaleadd/csa_gen.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (krenzland) [reply yateto/codegen/copyscaleadd/csa_gen.py:17]: Why don't we name the class CopyScaleAddGenerator or something? It's a bit more self-explanatory.

**Review by krenzland — CHANGES_REQUESTED** [review]
Looks mostly fine. I'd only like to rename CsaGenerator a clearer name - not a fan of (non widely used) abbreviations.
In the future, we need to take more care with regards to the generator and OOP design - the isinstance stuff is not great - but out of scope here.

**Review by uphoffc — APPROVED** [review]

**Review by krenzland — APPROVED** [review]
